include/aikido/planner/ompl/detail/OMPLConfigurationToConfigurationPlanner-impl.hpp

```diff
@@ -0,0 +1,207 @@
+
+  if (!boundsProjector)
+    boundsProjector
+        = std::move(createProjectableBounds(metaskeletonStatespace));
+
+  // Geometric State space
+  auto sspace = ompl_make_shared<GeometricStateSpace>(
+      mStateSpace,
+      std::make_shared<aikido::statespace::GeodesicInterpolator>(mStateSpace),
+      std::move(dmetric),
+      std::move(sampler),
+      mBoundsConstraint,
+      std::move(boundsProjector));
+
+  // Space Information
+  auto si = ompl_make_shared<::ompl::base::SpaceInformation>(std::move(sspace));
+
+  // Setup the planner
+  mPlanner = ompl_make_shared<PlannerType>(si);
+}
+
+//==============================================================================
+template <class PlannerType>
+trajectory::TrajectoryPtr
+OMPLConfigurationToConfigurationPlanner<PlannerType>::plan(
+    const SolvableProblem& problem, Result* result)
+{
+  auto si = mPlanner->getSpaceInformation();
+
+  // Set validity checker
+  std::vector<constraint::ConstTestablePtr> constraints{problem.getConstraint(),
+                                                        mBoundsConstraint};
+  auto conjunctionConstraint
+      = std::make_shared<constraint::TestableIntersection>(
+          mStateSpace, std::move(constraints));
+  ::ompl::base::StateValidityCheckerPtr vchecker
+      = ompl_make_shared<StateValidityChecker>(si, conjunctionConstraint);
+  si->setStateValidityChecker(vchecker);
+
+  ::ompl::base::MotionValidatorPtr mvalidator
+      = ompl_make_shared<MotionValidator>(si, mMaxDistanceBtwValidityChecks);
+  si->setMotionValidator(mvalidator);
+
+  // Define the OMPL problem
+  auto pdef = ompl_make_shared<::ompl::base::ProblemDefinition>(si);
+
+  // Only geometric statespaces are supported.
+  assert(ompl_dynamic_pointer_cast<GeometricStateSpace>(si->getStateSpace()));
+
+  auto sspace
+      = ompl_static_pointer_cast<GeometricStateSpace>(si->getStateSpace());
+  auto start = sspace->allocState(problem.getStartState());
+  auto goal = sspace->allocState(problem.getGoalState());
+
+  // ProblemDefinition clones states and keeps them internally
+  pdef->setStartAndGoalStates(start, goal);
+
+  sspace->freeState(start);
+  sspace->freeState(goal);
+
+  // Plan
+  mPlanner->setProblemDefinition(pdef);
+  mPlanner->setup();
+  auto solved = mPlanner->solve(1);
+
+  if (solved)
+  {
+    auto returnTraj = std::make_shared<trajectory::Interpolated>(
+        mStateSpace, mInterpolator);
+
+    // Get the path
+    auto path = ompl_dynamic_pointer_cast<::ompl::geometric::PathGeometric>(
+        pdef->getSolutionPath());
+    if (!path)
+    {
+      throw std::invalid_argument(
+          "Path is not of type PathGeometric. Cannot convert to aikido "
+          "Trajectory.");
+    }
+
+    for (std::size_t idx = 0; idx < path->getStateCount(); ++idx)
+    {
+      // Only states belonging geometric statespaces are supported.
+      assert(
+          dynamic_cast<aikido::planner::ompl::GeometricStateSpace::StateType*>(
```

Comment:
Is it possible for a `PathGeometric` to have non-`GeometricStateSpace::StateType`? If so, is it possible for a `PathGeometric` to have a mixture of `StateType`s (i.e. does this `dynamic_cast` need to happen every time)?